### PR TITLE
V3 updates

### PR DESCRIPTION
**Pull request by kreed0422**
Adam, review carefully before merging.  I took your suggestion about the methods in AbstractPgCheckService and created new route components so as not to break existing applications:

LoadPageGuardService, AbstractPageGuardService, PageStateService, DefaultPageGuardService

The Abstract class now only has 2 methods to define, one to bypass guards, the other for the logic that will determine whether or not to continue to the next page.  The DefaultPageGuardService shows how to code.

I was able to set this up in the test-page-service branch in MSP (enrolment only) and the route-guards in BCP.

### CODE REVIEW

File: projects/common/lib/services/page-state.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+function addKeys( x: any ): {[key: string]: any} {
+  return {
+    fullpath: x.fullpath,
+    title: x.title
+  };
+}
+
+describe('PageStateService', () => {
```
Comment (adamcoard-maximus) [inline projects/common/lib/services/page-state.service.spec.ts]: Love these tests!  Ran them locally and looks good.

File: projects/common/lib/services/page-state.service.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            routeListConst: any,
+            fn?: (x: any) => {[key: string]: any} ): PageList[] {
+
+    if ( this.pageList && !this.pageList.length ) {
+
+      const routeConst = Object.keys( routeListConst ).map( x => routeListConst[x] );
+
+      let cnt = 0;
+      this.pageList = arr.filter((itm: any) => !itm.redirectTo)
+        .map((itm: any) => {
```
Comment (adamcoard-maximus) [inline projects/common/lib/services/page-state.service.ts]: You can delete your `cnt` variable.

The "map" method on an array has multiple parameters, one of them includes an index.  So you can just use that.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

In practice, I think you'd have to change line 30 to (code is untested):

```suggestion
        .map((itm: any, index: number) => {
```

File: projects/common/lib/services/page-state.service.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          const val = routeConst.find( x => x.path === itm.path );
+
+          cnt = cnt + 1; // count for indices
+
+          // Page List to be returned
+          let obj = {
+            index: cnt,
```
Comment (adamcoard-maximus) [inline projects/common/lib/services/page-state.service.ts]: Assuming you name the variable "index", you can just use this syntactic sugar below:

```suggestion
            index,  
```

File: projects/common/lib/services/page-state.service.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            path: val.path,
+            isComplete: false,
+          };
+
+          if ( fn ) {
+            obj = Object.assign( obj, fn( val ) );
```
Comment (adamcoard-maximus) [inline projects/common/lib/services/page-state.service.ts]: Perfect!  I love this, such an elegant solution to a tricky problem. Functional languages are great.

**Comment by kreed0422** [comment]
I will update with suggestions.

**Pull request merged**